lib/saputils.pm

```diff
@@ -81,4 +82,87 @@ sub calculate_hana_topology {
     return \%topology;
 }
 
+=head2 check_hana_topology
+    check_hana_topology([input => $saphanasr_showAttr_format_script_output]);
+
+    Expect `SAPHanaSR-showAttr --format=script` as input.
+    Uses calculate_hana_topology to get a hash of hashes, and then
+    checks the output to make sure that the cluster is working and ready.
+
+    The checks performed are:
+    - All node_states are online
+    - All sync_states are either SOK or PRIM
+
+=cut
+
+
+sub check_hana_topology {
+    my (%args) = @_;
+    croak "Missing mandatory 'input' argument" unless $args{input};
+    record_info("cmd_out", $args{input});
+    my $topology = calculate_hana_topology(input => $args{input});
+
+    my $all_online = 1;
+    my $prim_count = 0;
+    my $sok_count = 0;
+
+    foreach my $host (keys %$topology) {
+        # Check node_state
+        unless ($topology->{$host}->{node_state} eq 'online') {
+            $all_online = 0;
+            last;
+        }
+
+        # Check sync_state
+        if ($topology->{$host}->{sync_state} eq 'PRIM') {
+            $prim_count++;
+        } elsif ($topology->{$host}->{sync_state} eq 'SOK') {
+            $sok_count++;
+        }
+    }
+
+    # Final check for conditions
+    if ($all_online && $prim_count == 1 && $sok_count == (keys %$topology) - 1) {
+        return 1;
+    } else {
+        return 0;
+    }    
+}
+
+=head2 check_crm_output
+    check_crm_output([input => $crm_mon_output]);
+
+    input: the output of the command 'crm_mon -r -R -n -N -1'
+    output: whether the conditions are met (return 1) or not (return 0)
+
+    Conditions:
+    - No resources are in 'Starting' state
+    - No 'Failed Resource Actions' present   
+
+=cut
+
+sub check_crm_output {
```

Comment:
Not for this PR but like an idea to further explore this design idea to split parsing and decision, I'd like to suggest to try to split this new function in  something like (pseudo, buggy wrong perl code):


```

sub parsing_crm_output {
    my (%args) = @_;
    croak "Missing mandatory 'input' argument" unless $args{input};
    record_info("cmd_out", $args{input});
    my %data;

    foreach $line in split '\n', $args{input} {
       if ($line =~ /:\s*Starting/) {
         $data{starting}++;
       }
       if ($line =~ /:\s*Started/) {
          $data{started}++;
       }
       # Check for Failed resource actions
       if ($line =~ /Failed Resource Actions:/) {
         $data{failed_actions}++;
       }
   }
  return \%data;
}

sub check_crm_output {
    my (%args) = @_;
    croak "Missing mandatory 'data' argument" unless $args{data};
    
    my $resource_started = 1;
    my $failed_actions = 0;
    # Check for resources that are still starting
    if ($args{input} =~ /:\s*Starting/) {
        $resource_started = 0;
    }
    # Check for Failed resource actions
    if ($args{input} =~ /Failed Resource Actions:/) {
        $failed_actions = 1;
    }
    if ($args{data}{resource_started} > 0 && !$args{data}{failed_actions}) {
        return 1;    
    } else {
        return 0;
    }
}```